### PR TITLE
Integrate PDP with OED as external authorization source

### PR DESCRIPTION
**Pull request by jonkjetiloye**
## Description
- Added new OedAuthzClient for integration with the oed-authz API
- Added new dependency to Altinn.ApiClients.Maskinporten v9.0.0
- Added new OedAuthzMaskinportenClientDefinition and Configuration
- Added new OedRoleAssignmentWrapper service
- Added new SBL bridge API client and service for getting SSN from UserId and PartyId

Other relevant changes:
- ContextHandler will now retrieve and evaluate policy for the App or Resource and evaluate the subjects of the rules. If any 'urn:altinn:rolecode' subjects are found Altinn roles will be retrieved. If any 'urn:oed:rolecode' subjects are found OED roles will be retrieved.

Additional PR made for studio-ops repo for both new (Maskinporten client config) and changes existing to configuration values (SBL bridge API path)

## Related Issue(s)
- #343

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

**Review by lovoll — APPROVED** [review]

File: src/Authorization/Models/Oed/OedRoleAssignmentRequest.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+namespace Altinn.Platform.Authorization.Models
+{
+    /// <summary>
+    /// Model for Access group membership
+    /// </summary>
+    public class OedRoleAssignmentRequest
+    {
+        /// <summary>
+        /// The OED role code
```
[the remark this answers is not in the document]
Comment (jonkjetiloye) [reply src/Authorization/Models/Oed/OedRoleAssignmentRequest.cs]: fixed

File: src/Authorization/Models/Oed/OedRoleAssignmentRequest.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    /// </summary>
+    public class OedRoleAssignmentRequest
+    {
+        /// <summary>
+        /// The OED role code 
+        /// </summary>
+        public string From { get; set; }
+
+        /// <summary>
+        /// The OED role code
```
[the remark this answers is not in the document]
Comment (jonkjetiloye) [reply src/Authorization/Models/Oed/OedRoleAssignmentRequest.cs]: fixed

File: src/Authorization/Clients/OedAuthzMaskinportenClientDefinition.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        /// </summary>
+        /// <param name="clientSettings">Maskinporten client settings</param>
+        public OedAuthzMaskinportenClientDefinition(IOptions<OedAuthzMaskinportenClientSettings> clientSettings) => ClientSettings = clientSettings.Value;
+
+        /// <inheritdoc/>
+        public Task<ClientSecrets> GetClientSecrets()
+        {
+            ClientSecrets clientSecrets = new ClientSecrets();
+
+            byte[] base64EncodedBytes = Convert.FromBase64String(ClientSettings.EncodedJwk);
```
[the remark this answers is not in the document]
Comment (jonkjetiloye) [reply src/Authorization/Clients/OedAuthzMaskinportenClientDefinition.cs]: renamed variable

File: src/Authorization/Helpers/PolicyHelper.cs [hunk 4]
```diff
@@ -51,6 +51,38 @@ public static List<string> GetRolesWithAccess(XacmlPolicy policy)
             return roleCodes.ToList();
         }
 
+        /// <summary>
+        /// Returns a list of uniue subject attribute ids. Used to identify the role/group and other sources needed to enirch subject info
+        /// </summary>
+        /// <param name="policy">The policy</param>
+        /// <returns>List of attribute ids</returns>
+        public static List<string> GetSubjectAttributeIds(XacmlPolicy policy)
```
Comment (TheTechArch) [inline src/Authorization/Helpers/PolicyHelper.cs:59]: A possible improvement could be to cache this list together with the policy. I am not sure if this iteration might be costly if it happens enough

File: src/Authorization/Services/Implementation/ProfileWrapper.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        /// <param name="profileClient">the client handler for profile api in Bridge</param>
+        /// <param name="logger">The logger</param>
+        public ProfileWrapper(ProfileClient profileClient, ILogger<ProfileWrapper> logger)
+        {
+            _profileClient = profileClient;
+            _logger = logger;
+        }
+
+        /// <inheritdoc/>
+        public async Task<UserProfile> GetUserProfile(int userId)
```
Comment (TheTechArch) [inline src/Authorization/Services/Implementation/ProfileWrapper.cs:36]: Is this something that could be cached?

File: src/Authorization/Helpers/PolicyHelper.cs (hunk 4 above)
Comment (jonkjetiloye) [reply src/Authorization/Helpers/PolicyHelper.cs:59]: Good point! Will look into it

File: src/Authorization/Services/Implementation/ProfileWrapper.cs (hunk 5 above)
Comment (jonkjetiloye) [reply src/Authorization/Services/Implementation/ProfileWrapper.cs:36]: It's currently being cached in the ContextHandler.GetSsnForUser where also the role, oedRole, keyRoleUnit caching etc. is located

Comment (jonkjetiloye) [reply src/Authorization/Services/Implementation/ProfileWrapper.cs:36]: I do agree though that the caching should be the responsibility of the service/wrapper implementation and not the consumer

File: src/Authorization/Helpers/PolicyHelper.cs (hunk 4 above)
Comment (jonkjetiloye) [reply src/Authorization/Helpers/PolicyHelper.cs:59]: Moving the implementation to the PRP and caching it with the policy feels like the wrong place.. and will need changes to the interface and existing use of the operations.

Ended up wrapping the use of the operation from ContextHandler in a GetFromCacheOrPolicy helper method.
But it then also required getting a PolicyId to be able to cache the result..

Comment (jonkjetiloye) [reply src/Authorization/Helpers/PolicyHelper.cs]: Moved as a helper function on the XacmlPolicy implementation in ABAC to extract all attributes for a given category

**Pull request merged**